objects/src/notes/stub.rs

```diff
@@ -0,0 +1,106 @@
+use super::{Digest, Felt, Hasher, NoteError, NoteMetadata, NoteVault, WORD_SIZE};
+use crate::parse_word;
+use crypto::StarkField;
+use miden_lib::memory::{
+    CREATED_NOTE_ASSETS_OFFSET, CREATED_NOTE_CORE_DATA_SIZE, CREATED_NOTE_HASH_OFFSET,
+    CREATED_NOTE_METADATA_OFFSET, CREATED_NOTE_RECIPIENT_OFFSET, CREATED_NOTE_VAULT_HASH_OFFSET,
+};
+
+// NOTE STUB
+// ================================================================================================
+
+/// An object that represents the stub of a note. When a note is produced in a transaction it can
+/// be the case that only the recipient, vault and metadata are known. In this case, the note
+/// stub can be used to represent the note.
+#[derive(Debug, Clone, PartialEq)]
+pub struct NoteStub {
+    hash: Digest,
+    recipient: Digest,
+    vault: NoteVault,
+    metadata: NoteMetadata,
+}
+
+impl NoteStub {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Creates a new [NoteStub].
+    pub fn new(
+        recipient: Digest,
+        vault: NoteVault,
+        metadata: NoteMetadata,
+    ) -> Result<Self, NoteError> {
+        if vault.num_assets() as u64 != metadata.num_assets().as_int() {
+            return Err(NoteError::InconsistentStubNumAssets(
+                vault.num_assets() as u64,
+                metadata.num_assets().as_int(),
+            ));
+        }
+        let hash = Hasher::merge(&[recipient, vault.hash()]);
+        Ok(Self {
+            hash,
+            recipient,
+            vault,
+            metadata,
+        })
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+    /// Returns the recipient of the note.
+    pub fn recipient(&self) -> &Digest {
+        &self.recipient
+    }
+
+    /// Returns a reference to the asset vault of this note.
+    pub fn vault(&self) -> &NoteVault {
+        &self.vault
+    }
+
+    /// Returns the metadata associated with this note.
+    pub fn metadata(&self) -> &NoteMetadata {
+        &self.metadata
+    }
+
+    /// Returns the hash of this note stub.
+    pub fn hash(&self) -> &Digest {
+        &self.hash
+    }
+}
+
+impl TryFrom<&[Felt]> for NoteStub {
+    type Error = NoteError;
+
+    fn try_from(elements: &[Felt]) -> Result<Self, Self::Error> {
+        if elements.len() < CREATED_NOTE_CORE_DATA_SIZE * WORD_SIZE {
+            return Err(NoteError::InvalidStubDataLen(elements.len()));
+        }
+
+        let hash: Digest = parse_word(elements, CREATED_NOTE_HASH_OFFSET).into();
```

Comment:
If we want to put this in `miden-lib` then we need to move everything at once as if we want this to be in `miden-lib` then we need to add `miden-objects` as a dependency to `miden-lib`, this would introduce a circular dependency.